### PR TITLE
Improvements to workflow & functionality for compiling scripts to C++

### PR DESCRIPTION
**Pull request by vpax**
This PR has two main elements:

1. An improved workflow for the common case of compiling scripts to C++.  The current workflow requires renaming files and creating empty ones.  The new workflow looks like:
> cd build
> src/zeek -O gen-C++ myscript.zeek
> ninja   (or make)
> src/zeek -O use-C++ myscript .zeek

2. Improved `-O gen-standalone-C++` functionality (embed scripts a C++ in a new `zeek` binary such that they can be used without having to provide the scripts in the future).  Previously, only the event handlers in standalone scripts would be activated.  Now, any types or functions that they define will be accessible, too.

There are also a couple of low-level tweaks: removal of dead code, and a fix for the availability of the `load_CPP()` BiF.

### CODE REVIEW

**Comment by vpax** [comment]
I've converted this to a Draft because @anthonykasza in alpha testing found a basic problem that I need to fix.

**Comment by vpax** [comment]
I fixed the issues that Anthony's testing found, so this is now ready again.

File: src/Func.cc [hunk 1]
```diff
@@ -332,8 +332,11 @@ ScriptFunc::ScriptFunc(std::string _name, FuncTypePtr ft,
 
 	sort(bodies.begin(), bodies.end());
 
-	current_body = bodies[0].stmts;
-	current_priority = bodies[0].priority;
+	if ( bodies.size() > 0 )
```
Comment (timwoj) [inline src/Func.cc:335]:
```suggestion
	if ( ! bodies.empty() )
```

File: src/Func.cc [hunk 2]
```diff
@@ -579,15 +582,22 @@ void ScriptFunc::ReplaceBody(const StmtPtr& old_body, StmtPtr new_body)
...
+				}
+			else
+				bodies.erase(body);
+
 			found_it = true;
+			break;
 			}
 
-	ASSERT(found_it);
+	// ASSERT(found_it);
```
Comment (timwoj) [inline src/Func.cc:600]: Can this be removed?

File: src/script_opt/CPP/Inits.cc [hunk 3]
```diff
@@ -548,7 +564,15 @@ void CPPCompile::GenLoad()
 
 	Emit("register_scripts__CPP(%s, standalone_init__CPP);", Fmt(total_hash));
 
-	// Spit out the placeholder script.
+	// Spit out the placeholder script, and any associated module
+	// definitions.
+	for ( auto& m : module_names )
```
Comment (timwoj) [inline src/script_opt/CPP/Inits.cc:569]: Prefer `const auto&` whenever it doesn't need to be modified.

File: src/script_opt/CPP/RuntimeInit.h [hunk 4]
```diff
@@ -20,6 +20,15 @@ typedef void (*CPP_init_func)();
 // Tracks the initialization hooks for different compilation runs.
 extern std::vector<CPP_init_func> CPP_init_funcs;
```
Comment (timwoj) [inline src/script_opt/CPP/RuntimeInit.h:21]: Ideallly we'd stop using function pointers and start using `std::function` for similar things, but I realize there's a ton of those all over this code base. That might be a good big cleanup task for someone to pick up.

**Review by timwoj — CHANGES_REQUESTED** [review]
I didn't review for the actual functionality added here, just from a C++-usage standpoint. Not a whole lot to comment on to be honest.

File: src/Func.cc (hunk 1 above)
Comment (vpax) [reply src/Func.cc]: Fixed in the commit I just pushed.

File: src/Func.cc (hunk 2 above)
Comment (vpax) [reply src/Func.cc]: Ditto.

File: src/script_opt/CPP/Inits.cc (hunk 3 above)
Comment (vpax) [reply src/script_opt/CPP/Inits.cc]: Ditto.

File: src/script_opt/CPP/RuntimeInit.h (hunk 4 above)
Comment (vpax) [reply src/script_opt/CPP/RuntimeInit.h:21]: Not an idiom on my radar, so I'll indeed leave it to others at some point in the future ...

Comment (timwoj) [reply src/script_opt/CPP/RuntimeInit.h:21]: Yep, I'll write an issue for it so I don't forget about it.

**Review by timwoj — APPROVED** [review]

**Pull request merged**